project_status_report/models/project_status_indicator.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Camptocamp (<http://www.camptocamp.com>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+from odoo import api, fields, models
+from odoo import _, exceptions
+from odoo import SUPERUSER_ID
+
+import sys
+import compiler
+import traceback
+
+from .common import PYTHON_CODE_DEFAULT
+
+import logging
+logger = logging.getLogger('project_status_indicator')
+
+
+class ProjectStatusIndicator(models.Model):
+    _name = 'project.status.indicator'
+
+    _order = 'sequence'
+
+    name = fields.Char(required=True, translate=True)
+    report_id = fields.Many2one(comodel_name='project.status.report')
+    sequence = fields.Integer(default=10)
+    value_type = fields.Selection(selection=[('numeric', _('Numeric')),
+                                             ('boolean', _('Boolean')),
+                                             ('text', _('Text'))
+                                             ],
+                                  required=True,
+                                  )
+    python_code = fields.Text(default=PYTHON_CODE_DEFAULT)
+
+    @api.multi
+    @api.constrains('python_code')
+    def _check_python_code_syntax(self):
+        """
+        Syntax check the python code of a step
+        """
+        for step in self:
+            try:
+                compiler.parse(step.python_code)
```

Comment:
:hand: you should use safe_eval. [Example](https://github.com/camptocamp/odoo-prototype/blob/10.0/base_dj/models/dj_song.py#L195).

Reply:
Here you should use safe_eval
https://github.com/odoo/odoo/blob/10.0/odoo/tools/safe_eval.py

You can do:

``` python
safe_eval(self.python_code.strip(), ld,
          mode="exec", nocopy=True)
ld.get('color', None)
ld.get('value', None)
```